test/unit/ngsiv2/general/deviceService-test.js

```diff
@@ -0,0 +1,265 @@
+                            name: 'status',
+                            type: 'Boolean'
+                        }
+                    ],
+                    static_attributes: []
+                }
+            ]
+        },
+        headers: {
+            'fiware-service': 'TestService',
+            'fiware-servicepath': '/testingPath'
+        }
+    },
+    deviceCreation = {
+        url: 'http://localhost:' + iotAgentConfig.server.port + '/iot/devices',
+        method: 'POST',
+        json: utils.readExampleFile('./test/unit/examples/deviceProvisioningRequests/provisionNewDevice.json'),
+        headers: {
+            'fiware-service': 'TestService',
+            'fiware-servicepath': '/testingPath'
+        }
+    },
+    contextBrokerMock,
+    iotamMock;
+
+
+/* jshint camelcase: false */
+describe('Device Service: utils', function() {
+    beforeEach(function(done) {
+        nock.cleanAll();
+        logger.setLevel('FATAL');
+        iotamMock = nock('http://localhost:8082')
+            .post('/protocols')
+            .reply(200, {});
+
+        iotAgentLib.activate(iotAgentConfig, done);
+    });
+
+    afterEach(function(done) {
+        nock.cleanAll();
+        async.series([
+            iotAgentLib.clearAll,
+            iotAgentLib.deactivate
+        ], done);
+    });
+
+    describe('When an existing device tries to be retrieved with retrieveOrCreate()', function() {
+        beforeEach(function(done) {
+
+            // This mock does not check the payload since the aim of the test is not to verify 
+            // device provisioning functionality. Appropriate verification is done in tests under
+            // provisioning folder
+            contextBrokerMock = nock('http://192.168.1.1:1026')
+                .matchHeader('fiware-service', 'TestService')
+                .matchHeader('fiware-servicepath', '/testingPath')
+                .post('/v2/entities?options=upsert')
+                .reply(204);
+
+            async.series([
+                request.bind(request, groupCreation),
+                request.bind(request, deviceCreation)
+            ], function(error, results) {
+                done();
+            });
+        });
+
+        it('should return the existing device', function(done) {
+            iotAgentLib.retrieveDevice('Light1', '801230BJKL23Y9090DSFL123HJK09H324HV8732', function(error, device) {
+                should.not.exist(error);
+                should.exist(device);
+
+                device.id.should.equal('Light1');
+                done();
+            });
+        });
+    });
+
+    describe('When an unexisting device tries to be retrieved for an existing APIKey', function() {
+        beforeEach(function(done) {
+            contextBrokerMock = nock('http://192.168.1.1:1026')
+                .matchHeader('fiware-service', 'TestService')
+                .matchHeader('fiware-servicepath', '/testingPath')
+                .post('/v2/entities?options=upsert')
```

Comment:
Mock using post without payload. Comment explaining it needed?

Reply:
Fixed with fbc86fa76f4561b711a63053695efd21941cf4e5